📖 External Remediation Proposal (#3190)
kubernetes-sigs/cluster-api: docs/proposals/20200615-external-remediation.md

---

> title: external remediation

Reviewer: the title should start with uppercase?

         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
            * [Flow](#flow)
            * [Proposed CRD changes and additions](#proposed-crd-changes-and-additions)
            * [Example CRs](#example-crs)
            * [Usage](#usage)
         * [Risks and Mitigations](#risks-and-mitigations)
      * [Alternatives](#alternatives)
         * [Alternative 1 - Integrate power-cycling into the main MHC flow](#alternative-1---integrate-power-cycling-into-the-main-mhc-flow)
         * [Alternative 2 - Utilize conditions](#alternative-2---utilize-conditions)
         * [Alternative 3 - Create a dedicated MHC](#alternative-3---create-a-dedicated-mhc)
      * [Upgrade Strategy](#upgrade-strategy)
      * [Additional Details](#additional-details)
         * [Test Plan [optional]](#test-plan-optional)
         * [Graduation Criteria [optional]](#graduation-criteria-optional)
         * [Version Skew Strategy [optional]](#version-skew-strategy-optional)
      * [Implementation History](#implementation-history)

## Glossary

Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).

## Summary

> Enable opt in external remediation for unhealthy nodes backed by machines.

Reviewer: opt-in

Author: Changed

### Non-Goals/Future Work

- Provide a [unified remediation flow](https://docs.google.com/document/d/1klvIR5wKbcCrl2vLPIbbWpmelGTDM-T0Qb4xw36P3zY/edit) as part of the MachineHealthCheck controller. See Alternative 1
- Add power primitives to the Machine API.  While useful, it is unnecessary to the work proposed here and may already be covered by one or more of the following issues:
    - Add Ability to Reboot Hosts
    - Supporting power off/on
    - Add machine/infra feature support fields
- Alter the default deletion flow proposed in [cluster-api PR#3056](https://github.com/kubernetes-sigs/cluster-api/pull/3056)
- Provide a finalizer-like pre-hook mechanism for allowing customization or blocking of the remediation process prior to power-cycling a machine or deleting it. This concept may already be covered as part of a separate enhancement
- Create an escalation path from failed external remediation attempts to machine deletion.
- Implement a formal mechanism for triggering recovery of workload and volumes by the k8s scheduler

## Proposal

We introduce a generic mechanism for supporting externally provided custom remediation strategies.

> We propose modifying the MachineHealthCheck CRD to support a remediationTemplate, an ObjectReference to a provider-specific template CRD. 

Author: Hmm, I believe that this relies on `generateTemplate(..)` which receives an `ObjectReference`.
Can we use `generateTemplate()` without having an `ObjectReference`?

> We propose modifying the MachineHealthCheck CRD to support a remediationTemplate, an ObjectReference to a provider-specific template CRD. 

Reviewer 2: We could likely refactor the core logic of generateTemplate into a separate function that takes individual parameters (or an input struct) rather than an ObjectReference, calling that function from both net new code and still providing generateTemplate() for backward compatibility until we've removed existing use of ObjectReference.

> If no value for remediationTemplate is defined for the MachineHealthCheck CR, the existing annotation-based deletion flow is preserved.

Reviewer 3: ```suggestion
If no value for remediationTemplate is defined for the MachineHealthCheck CR, the existing condition-based deletion flow is preserved.
```

> If a value for remediationTemplate is supplied and the Machine enters an unhealthy state, the template will be instantiated using existing CAPI functionality, with the same name and namespace as the target Machine, and the remediation flow passed to an External Remediation Controller (ERC) watching for that CR.  

Author: As I see it, if it exists, it means that the Machine is unhealthy.
If the machine is healthy, it's MHC responsibility to delete that CR.

> Or maybe the potential for a name collision is actually a good thing and an indication that we shouldn't be creating a new resource, since there is likely a remediation operation already underway...

exactly

> No further action (deletion or applying conditions) will be taken by the MachineHealthCheck controller until the Node becomes healthy, when it will locate and delete the instantiated MachineRemediation CR.

Reviewer 4: we are currently working with Remediation Controller proposal in CAPM3. This is something we would like to see in place when external remediation proposal is approved. Is anyone against it that MHC will delete the MachineRemediation CR when Node becomes healthy.

Author: > Is the idea that MHC continues to perform health checks, just not attempts to mark the Machine for the "normal" remediation paths?

correct

> Does MachineHealthCheck wait for the machine remediation CR to report that it's finished prior to allowing for the deletion of the machine remediation CR? If not, I suspect we need to potentially worry about race conditions.

MHC doesn't wait. Can you provide examples for race conditions that may occur? we thought that ERC can add a finalizer on the machine remediation CR, if it needs to, and this might help to avoid race conditions

> No further action (deletion or applying annotations) will be taken by the MachineHealthCheck controller until the Node becomes healthy, when it will locate and delete the instantiated MachineRemediation CR.

Reviewer 3: ```suggestion
No further action (deletion or applying conditions) will be taken by the MachineHealthCheck controller until the Node becomes healthy, when it will locate and delete the instantiated MachineRemediation CR.
```

> The ERC may wish to register a finalizer on it’s CR to ensure it has an opportunity to perform any additional cleanup in the case that the unhealthy state was transient and the Node returned to a healthy state prior to the completion of the full custom ERC flow.

Reviewer 3: The nittiest of nits. My way of remembering this is "it's" should always be replaceable by "it is" or "it has"  😄 

```suggestion
The ERC may wish to register a finalizer on its CR to ensure it has an opportunity to perform any additional cleanup in the case that the unhealthy state was transient and the Node returned to a healthy state prior to the completion of the full custom ERC flow.
```

### User Stories

#### Story 1
As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can recover from transient errors faster and begin application recovery sooner.
#### Story 2

> As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can detect hardware issues faster.

Author: machine can go unhealthy due to software or hardware problems.
if you automatically power-cycle the host, it saves some time for an admin that would do exactly that to see if it was some temporary failure or something consistent.

but even if it's consistent, it still can be a software issue, so maybe this needs some clarification.
@beekhof - is this was the intention here?

> As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can detect hardware issues faster.

Author: discussed this with beekhof.
the intention was what I wrote in previous comment - to eliminate transient issues, either caused by software or hardware,
I'll rephrase to reflect that.
thanks

#### Story 3

> As an admin of a hardware based cluster, I would like the system to keep attempting to power-cycle unhealthy nodes, so that they are automatically added back to the cluster when I fix the underlying problem.

Author: > Can you provide an example of when this would be intended behavior? 

routing issues for example that are external to the server itself but prevents it to reach the cluster network

> why would we want to continually restart until a technician can fix the server rather than trying to provision on a different server?

That's a good point, and I think that in one of the discussion it was suggested to have some kind of max reboot attempts, but I think this is up to the external remediation controller to decide

#### Story 4

> As an admin of a large bare metal cluster, I want to auto remediate nodes but do so grouped by failure domain (i.e. by rack). All nodes deemed unhealthy in a given rack should be cordoned first before each are drained.

Author: I agree, I'll remove this one. thanks

### Implementation Details/Notes/Constraints

#### Flow

> The default MHC remediation strategy continues to be an annotation-based flow where the owning controller deletes the Machine (and by inference the Node) objects after performing any relevant cleanup.

Reviewer 3: ```suggestion
The default MHC remediation strategy continues to be an condition-based flow where the owning controller deletes the Machine (and by inference the Node) objects after performing any relevant cleanup.
```

In order to enable a different remediation strategy for a set of Machines, the admin must provide a reference to a provider-specific template object as part of the MachineHealthCheck CR.

When a Machine enters an unhealthy state, the MHC will:
* Look up the referenced template
* Instantiate the template (for simplicity, we will refer to this as a External Machine Remediation CR, or EMR)
* Force the name and namespace to match the unhealthy Machine
* Save the new object in etcd

We use the same name and namespace for the External Machine Remediation CR to ensure uniqueness and lessen the possibility for multiple parallel remediations of the same Machine. 

The lifespan of the EMRs is that of the remediation process and they are not intended to be a record of past events.  The EMR will also contain an ownerRef to the Machine, to ensure that it does not outlive the Machine it references.

> The only signaling between the MHC and the external controller watching EMR CRs is the creation and deletion of the EMR itself.  Any actions or changes that admin’s should be informed about should be emitted as events for consoles and UIs to consume if necessary.  They are informational only and do not result in or expect any behaviour from the MHC, Node, or Machine as a result.

Reviewer 3: We can also use conditions for this signaling, this is the new standard we've been moving to.

```suggestion
The only signaling between the MHC and the external controller watching EMR CRs is the creation and deletion of the EMR itself.  Any actions or changes that admins should be informed about should be emitted as events for consoles and UIs to consume if necessary.  They are informational only and do not result in or expect any behaviour from the MHC, Node, or Machine as a result.
```

When the external remediation controller detects the new EMR it starts remediation and performs whatever actions it deems appropriate until the EMR is deleted by the MHC. It is a detail of the ERC when and how to retry remediation in the event that a EMR is not deleted after the ERC considers remediation complete. 

The ERC may wish to register a finalizer on its CR to ensure it has an opportunity to perform any additional cleanup in the case that the unhealthy state was transient and the Node returned to a healthy state prior to the completion of the full custom ERC flow.

#### Proposed CRD changes and additions

    type MachineHealthCheckSpec struct {
        ...
    
        // +optional

>         RemediationTemplate ObjectReference `json:"remediationTemplate,omitempty"`

Reviewer 5: nit
go doc missing (It will help to clarify the intent of this doc)
Also, what about adding a prefix like "external" or "custom" in order to make it more explicit that this field is optional and to suggest that if the value is empty a default behavior applies?

Reviewer 4: "ExternalRemediationTemplate" could make sense.

Reviewer 2: This should also likely be a custom type scoped to the exact type of information needed, rather than corev1.ObjectReference.

Thinking aloud, I suspect the following information would be needed:
apigroup
kind
name

We should also likely avoid the use of storing the version for the referenced type, otherwise we potentially need to worry about migration strategies for these loosely coupled types.

However if we don't encode the version, then instantiating the template would require a way to lookup (through discovery or some other means) the correct version of the resource to create.

Author: > "ExternalRemediationTemplate" could make sense.

:+1: 
> This will need to be a pointer

:+1: 

> I suspect the following information would be needed:
apigroup
kind
name

isn't it what we have in MHC CR under `externalRemediationTemplate`?

    }

#### Example CRs

MachineHealthCheck:
    
    kind: MachineHealthCheck
    apiVersion: cluster.x-k8s.io/v1alphaX
    metadata:
      name: REMEDIATION_GROUP
      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
    spec:
      selector:
        matchLabels: 
          ...
      remediationTemplate:
        kind: Metal3RemediationTemplate
        apiVersion: remediation.metal3.io/v1alphaX
        name: M3_REMEDIATION_GROUP

Metal3RemediationTemplate:

    kind: Metal3RemediationTemplate
    apiVersion: remediation.metal3.io/v1alphaX
    metadata:
      name: M3_REMEDIATION_GROUP
      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
    spec:
      template:
        spec:
          strategy: 		    escalate
          deleteAfterRetries:     10
          powerOnTimeoutSeconds:  600
          powerOffTimeoutSeconds: 120

Metal3Remediation:

    apiVersion: remediation.metal3.io/v1alphaX
    kind: Metal3Remediation
    metadata:
      name: NAME_OF_UNHEALTHY_MACHINE
      namespace: NAMESPACE_OF_UNHEALTHY_MACHINE
       finalizer: 
       - remediation.metal3.io
       ownerReferences: 
    apiVersion:cluster.x-k8s.io/v1alphaX
    kind: MachineHealthCheck
    name: REMEDIATION_GROUP

>     uid: ...

Author: you're right. fixed

      spec: 
        strategy: 		    escalate
        deleteAfterRetries:     10
        powerOnTimeoutSeconds:  600
        powerOffTimeoutSeconds: 120
      status:
        phase: power off
        retryCount: 1

#### Usage

    
    func (t *healthCheckTarget) createProviderRemediation(ctx context.Context) (*corev1.ObjectReference, error) {

>         if t.MHC.Spec.RemediationTemplate == nil {

Reviewer 6: This should return an error if the template is nil

            return nil, nil
    }
    
    from, err := Get(ctx, c, t.MHC.Spec.RemediationTemplate, t.Machine.Namespace)
    if err != nil {
        return nil, err
    }
    generateTemplateInput := &GenerateTemplateInput{
        Template:    from,
        Namespace:   t.Machine.Namespace,
        ClusterName: t.Machine.ClusterName,
        OwnerRef:    metav1.OwnerReference{*metav1.NewControllerRef(
    t.Machine, t.Machine.Kind)},
        Labels:      [],
    }
    to, err := GenerateTemplate(generateTemplateInput)
    if err != nil {
        return nil, err
    }
    
    // Force the same name as the machine
    to.SetName(t.Machine.Name)
    
    // Create the external clone.
    if err := r.Client.Create(context.Background(), to); err != nil {
        return nil, err
    }
    return GetObjectReference(to), nil
    }
    
    func (t *healthCheckTarget) remediate(...) error {
        ...

>     remediation, err := t.createProviderRemediation(ctx)

Reviewer 6: Before I review the code suggestions any further, how would you feel about removing them from the proposal, and we'd defer reviewing the code until there is a PR?

Reviewer 4: That make sense.

Author: Sure, I'll remove the code. thanks

    if remediation != nil && err != nil {
        // control passes to external remediation controller
        return nil
    } else if err != nil && errors.IsAlreadyExists(err) {
        // external remediation controller already has control
    return nil
    } else if err != nil {
        logger.Errorf("%q: failed to create provider remediation object for Machine %q in namespace %q", err, t.machine.Name, t.machine.Namespace)

>     }

Reviewer 6: Please correct the indentation

    ...
    }

### Risks and Mitigations

> - Currently an admin can configure multiple MHCs with overlapping selectors, meaning that he can configure different remediation templates to the same Machine. This will currently be resolved by proper documentation as described in [issue #3058](https://github.com/kubernetes-sigs/cluster-api/issues/3058)

Reviewer 6: I would not consider this a risk that is created by or specific to this proposal. Want to remove it?
